### PR TITLE
discover nested vsconfigs

### PR DESCRIPTION
**Pull request by jlee671**
Discover nested vsconfigs and collect components to create final vsconfig

- [x] Manually tested discovering directories where .vsconfig exists and collecting components (unique)
paths
components
finalvsconfig

### CODE REVIEW

File: VSConfigFinder/FileSystem.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+            return Directory.GetFileSystemEntries(path, pattern, recursive ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly);
+        }
+
+        /// <inheritdoc/>
+        public Stream OpenFile(string path)
+        {
+            Utilities.ValidateIsNotNullOrEmpty(path, nameof(Path));
+
+            return File.OpenRead(path);
```
Comment (skylarnam) [inline VSConfigFinder/FileSystem.cs:27]: It'd be great to merge `CreateOutput` utility method where it calls `File.WriteAllText` into this new class. What do you think? (Even if it's just a one-liner plus some validation)

File: VSConfigFinder/FileSystem.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+﻿namespace VSConfigFinder
```
Comment (skylarnam) [inline VSConfigFinder/FileSystem.cs]: Add file header

File: VSConfigFinder/IFileSystem.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿namespace VSConfigFinder
```
Comment (skylarnam) [inline VSConfigFinder/IFileSystem.cs]: Add file header

File: VSConfigFinder/IFileSystem.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+{
+
+    public interface IFileSystem
+    {
+        public string[] GetFileSystemEntries(string path, string pattern, bool recursive = false);
```
Comment (skylarnam) [inline VSConfigFinder/IFileSystem.cs:22]: Add comments

File: VSConfigFinder/Program.cs [hunk 5]
```diff
@@ -38,17 +39,12 @@ private static void Run(CommandLineOptions options)
                 options.ConfigOutputPath ??= Directory.GetCurrentDirectory();
```
Comment (skylarnam) [inline VSConfigFinder/Program.cs]: Now that we have all the parts, should we pull out this if check then assign part to something like `ResolveOptions(CommandLineOptions options)`?

File: VSConfigFinder/Program.cs [hunk 6]
```diff
@@ -38,17 +39,12 @@ private static void Run(CommandLineOptions options)
...
             }
 
-            // TODO: Replace this with the output from the runner
+            string[] allComponents = Utilities.ReadComponents(options);
```
Comment (skylarnam) [inline VSConfigFinder/Program.cs]: use `var`

Comment (skylarnam) [reply VSConfigFinder/Program.cs]: Get rid of todo

File: VSConfigFinder/Utilities.cs [hunk 7]
```diff
@@ -70,5 +70,29 @@ private static string CreateCommandLineOutput(VSConfig finalConfig)
...
             return output.ToString();
         }
+
+        public static string[] ReadComponents(CommandLineOptions options)
+        {
+            var fileSystem = new FileSystem();
+            var pathsToVsConfigs = fileSystem.GetFileSystemEntries(options.FolderPath, ConfigExtension, true);
+
+            HashSet<string> componentsSet = new HashSet<string>();
+            var serializerOptions = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
```
Comment (skylarnam) [inline VSConfigFinder/Utilities.cs]: Are there other properties we need to set for some edge cases?

File: VSConfigFinder/Utilities.cs [hunk 8]
```diff
@@ -70,5 +70,29 @@ private static string CreateCommandLineOutput(VSConfig finalConfig)
 
             return output.ToString();
         }
+
+        public static string[] ReadComponents(CommandLineOptions options)
+        {
+            var fileSystem = new FileSystem();
+            var pathsToVsConfigs = fileSystem.GetFileSystemEntries(options.FolderPath, ConfigExtension, true);
+
+            HashSet<string> componentsSet = new HashSet<string>();
```
Comment (skylarnam) [inline VSConfigFinder/Utilities.cs]: use var

File: VSConfigFinder/Utilities.cs [hunk 9]
```diff
@@ -70,5 +70,29 @@ private static string CreateCommandLineOutput(VSConfig finalConfig)
 
             return output.ToString();
         }
+
+        public static string[] ReadComponents(CommandLineOptions options)
+        {
+            var fileSystem = new FileSystem();
+            var pathsToVsConfigs = fileSystem.GetFileSystemEntries(options.FolderPath, ConfigExtension, true);
```
Comment (skylarnam) [inline VSConfigFinder/Utilities.cs]: nit: use argument name for flags (true/false)

File: VSConfigFinder/Utilities.cs [hunk 10]
```diff
@@ -70,5 +70,29 @@ private static string CreateCommandLineOutput(VSConfig finalConfig)
...
+
+            HashSet<string> componentsSet = new HashSet<string>();
+            var serializerOptions = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
+
+            foreach (var path in pathsToVsConfigs)
+            {
+                var stream = fileSystem.OpenFile(path);
+                var components = JsonSerializer.Deserialize<VSConfig>(stream, serializerOptions).Components;
+
+                if (components != null)
```
Comment (skylarnam) [inline VSConfigFinder/Utilities.cs]: nit: is not

File: VSConfigFinder/Utilities.cs [hunk 11]
```diff
@@ -70,5 +70,29 @@ private static string CreateCommandLineOutput(VSConfig finalConfig)
...
+            var fileSystem = new FileSystem();
+            var pathsToVsConfigs = fileSystem.GetFileSystemEntries(options.FolderPath, ConfigExtension, true);
+
+            HashSet<string> componentsSet = new HashSet<string>();
+            var serializerOptions = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
+
+            foreach (var path in pathsToVsConfigs)
+            {
+                var stream = fileSystem.OpenFile(path);
+                var components = JsonSerializer.Deserialize<VSConfig>(stream, serializerOptions).Components;
```
Comment (skylarnam) [inline VSConfigFinder/Utilities.cs]: Should we wrap this around so that we can deal with exceptions? + Do a null check on the deserialized output

File: VSConfigFinder/Program.cs [hunk 12]
```diff
@@ -7,6 +7,7 @@ namespace VSConfigFinder
 {
```
Comment (skylarnam) [inline VSConfigFinder/Program.cs]: Add unit tests to the changes

**Review by skylarnam — CHANGES_REQUESTED** [review]

File: VSConfigFinder/FileSystem.cs (hunk 1 above)
Comment (jlee671) [reply VSConfigFinder/FileSystem.cs:27]: Like in FileSystem.cs, create method that does the if block logic in CreateOutput? If so, I think it makes sense since that block does File stuff.

File: VSConfigFinder/Utilities.cs (hunk 11 above)
Comment (jlee671) [reply VSConfigFinder/Utilities.cs]: Yea null check is done below, and both OpenFile and Deserialize can throw exception, I will write try catch block. When we catch it, should we simply Console.WriteLine to let user know what happened?

File: VSConfigFinder/Utilities.cs (hunk 7 above)
Comment (jlee671) [reply VSConfigFinder/Utilities.cs]: I was browsing through the options, IgnoreNullValues and AllowTrailingCommas could be helpful. What do you think?

File: VSConfigFinder/FileSystem.cs (hunk 1 above)
Comment (skylarnam) [reply VSConfigFinder/FileSystem.cs:27]: Yep, if you look at our Setup code we wrap all the `File.xx` calls with the FileSystem calls so that it can be testable with mocks. If you could do that while you're on it, it'd be great!

File: VSConfigFinder/Utilities.cs (hunk 7 above)
Comment (skylarnam) [reply VSConfigFinder/Utilities.cs]: You can test a few edge cases yourself, but I think that could be useful.

File: VSConfigFinder/Utilities.cs (hunk 11 above)
Comment (skylarnam) [reply VSConfigFinder/Utilities.cs]: For example, `JsonSerializer.Deserialize<T>` returns an `object?` which means that the output could be null. If we don't do a null check here in any way, doing a `.Components` after that could throw.

Upon catch, I'd say we write a log through `Console.WriteLine` to signal the user that it failed & why it failed.

**Review by skylarnam — APPROVED** [review]

File: VSConfigFinder/Program.cs [hunk 13]
```diff
@@ -23,36 +23,34 @@ public static void Main(string[] args)
...
-                    "Microsoft.VisualStudio.Component.Roslyn.Compiler",
-                    "Microsoft.Component.MSBuild",
-                    "Microsoft.NetCore.Component.Runtime.6.0",
-                },
+                Components = Utilities.ReadComponents(fileSystem, options),
             };
 
-            // Output
-            Utilities.CreateOutput(finalConfig, options);
+            Utilities.CreateOutput(fileSystem, finalConfig, options);
```
Comment (tydunkel) [inline VSConfigFinder/Program.cs:45]: nit: why not just pass the components here and generate the vsconfig in here, if needed based on the cmdline switches?

**Review by tydunkel — APPROVED** [review]

File: VSConfigFinder/Program.cs (hunk 13 above)
Comment (skylarnam) [reply VSConfigFinder/Program.cs:45]: Mostly just for testability, I wanted to move most of the logic out of `program.cs`.
